### PR TITLE
Use LazySeriesIterator with fuzzy metric name queries

### PR DESCRIPTION
**Pull request by aaron7**
Step towards fixing #416

**Create v8Schema with series index (#430)**

This index will be used to fetch series from the index. It will replace the metric name index and we will make the switch after the next steps are complete.

Decided to encode the model.Metric as JSON because it already implements the json.Unmarshaler interface.

Used sha256 for identifying the series as the variant of fnv64a which if used for fingerprints is not unique enough for indexing and using sha256 keeps it simple (we don't have to deal with collision logic).

**Move iterators inside chunk store using MergeSeriesIterator (#438)**

- Move iterators inside the chunk store
- Create MergeSeriesIterator which is created from multiple iterators and implements SeriesIterator.
- Move iterator code to util as it is used across packages and cyclic imports would exist in some cases if it was kept in querier.

**Use LazySeriesIterator with fuzzy metric name queries**

There is not opportunity for our iterator to return an error if values are request, so we are just returning no values for now.

Test plan:
- `count({__name__=~".+"}) by (__name__)` ran locally

### CODE REVIEW

File: pkg/chunk/chunk_store.go [hunk 1]
```diff
@@ -151,17 +152,35 @@ func (c *Store) calculateDynamoWrites(userID string, chunks []Chunk) (WriteBatch
 }
 
 // Get implements ChunkStore
-func (c *Store) Get(ctx context.Context, from, through model.Time, allMatchers ...*metric.LabelMatcher) ([]Chunk, error) {
+func (c *Store) Get(ctx context.Context, from, through model.Time, allMatchers ...*metric.LabelMatcher) ([]local.SeriesIterator, error) {
 	if through < from {
 		return nil, fmt.Errorf("invalid query, through < from (%d < %d)", through, from)
 	}
 
 	filters, matchers := util.SplitFiltersAndMatchers(allMatchers)
```
Comment (tomwilkie) [inline pkg/chunk/chunk_store.go]: I wonder if this should be pushed into `getMetricNameIterators` and `getFuzzyMetricLazySeriesIterators`.

File: pkg/chunk/chunk_store.go [hunk 2]
```diff
@@ -209,69 +228,52 @@ outer:
...
-func (c *Store) lookupChunksByMatchers(ctx context.Context, from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
-	metricNameMatcher, matchers, ok := util.ExtractMetricNameMatcherFromMatchers(matchers)
-
-	// Only lookup by metric name if the matcher is of type equal, otherwise we
-	// have to fetch chunks for all metric names as other metric names could match.
-	if ok && metricNameMatcher.Type == metric.Equal {
-		return c.lookupChunksByMetricName(ctx, from, through, matchers, metricNameMatcher.Value)
-	}
-
+func (c *Store) getFuzzyMetricLazySeriesIterators(ctx context.Context, from, through model.Time, filters []*metric.LabelMatcher, matchers []*metric.LabelMatcher, metricNameMatcher *metric.LabelMatcher) ([]local.SeriesIterator, error) {
```
Comment (tomwilkie) [inline pkg/chunk/chunk_store.go]: Probably better names `getSeriesIterators`?

File: pkg/chunk/chunk_store.go [hunk 3]
```diff
@@ -209,69 +228,52 @@ outer:
...
 		if err != nil {
 			return nil, err
 		}
 
-		// We are fetching all metric name chunks, however if there is a metricNameMatcher,
-		// we only want metric names that match
-		if ok && !metricNameMatcher.Match(metricName) {
-			skippedMetricNames++
+		// Apply metricNameMatcher filter
+		if metricNameMatcher != nil && !metricNameMatcher.Match(metric[metricNameMatcher.Name]) {
```
Comment (tomwilkie) [inline pkg/chunk/chunk_store.go]: Is this not redundant with the checks below?

Also, if you push the `util.SplitFiltersAndMatchers` into `getMetricNameIterators`, this section will just become:

```
for _, matcher := range matchers {
    if !... {...}
}
```

And you won't need all three checks.

File: pkg/chunk/schema_test.go [hunk 4]
```diff
@@ -236,6 +237,7 @@ func TestSchemaHashKeys(t *testing.T) {
 const (
 	MetricNameRangeValue = iota + 1
```
Comment (tomwilkie) [inline pkg/chunk/schema_test.go]: typically you do 

```go
const (
   _ iota
   foo
   bar
)
```

If you want to start at 1.

File: pkg/chunk/schema_util_test.go [hunk 5]
```diff
@@ -11,6 +13,34 @@ import (
...
+		},
+		{
+			model.Metric{
+				model.MetricNameLabel: "foo",
+				"bar":  "baz",
+				"toms": "code",
+				"flip": "flop",
+			},
+			"KrbXMezYneba+o7wfEdtzOdAWhbfWcDrlVfs1uOCX3M",
+		},
```
Comment (tomwilkie) [inline pkg/chunk/schema_util_test.go]: Can you (for my sanity) add a test case for the same metric, with the pairs in a different order?

File: pkg/distributor/distributor.go [hunk 6]
```diff
@@ -517,7 +532,7 @@ func (d *Distributor) queryIngesters(ctx context.Context, ingesters []*ring.Inge
 					}
 					fpToSampleStream[fp] = mss
 				}
-				mss.Values = util.MergeSamples(mss.Values, ss.Values)
+				mss.Values = util.MergeSampleSets(mss.Values, ss.Values)
```
Comment (tomwilkie) [inline pkg/distributor/distributor.go]: In such large PRs I generally try and avoid renamings.

File: pkg/util/iterator.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
+package util
```
Comment (tomwilkie) [inline pkg/util/iterator.go]: Add a comment that this is copied from the prometheus code, and the PR to make it public?

**Comment by tomwilkie** [comment]
Let close #430 and #438 and put all the changes in one PR (this one).  That way we won't miss any comments.

Done the first pass - looks good!  Not obvious errors, just a bunch of nits.

Have you tested locally?  I didn't see any changes to the checked-in k8s config to enable this.

File: pkg/distributor/distributor.go (hunk 6 above)
Comment (aaron7) [inline pkg/distributor/distributor.go]: We changed it to `MergeSampleSets` to keep it consistent with `MergeNSampleSets` and it's only used in 2 places. I believe this was in a previous PR but we have now agreed to merge all PR's together.

**Comment by tomwilkie** [comment]
Is this gtg now? I'll review this weekend if so.

File: pkg/chunk/schema.go [hunk 8]
```diff
@@ -529,6 +539,39 @@ func (entries v7Entries) GetWriteEntries(bucket Bucket, metricName model.LabelVa
 }
 
 func (v7Entries) GetReadQueries(bucket Bucket) ([]IndexQuery, error) {
+	// Replaced with v8Schema series index
+	return nil, ErrNoMetricNameNotSupported
+}
+
+// v8Entries is the same as v7Entries however with a series index instead of a metric name index
```
Comment (marccarre) [inline pkg/chunk/schema.go]: `v8Entries is the same as v7Entries`
does not seem consistent with:
`type v8Entries struct { v6Entries }`
which one is correct? is just a typo in the comment?

Comment (aaron7) [inline pkg/chunk/schema.go]: `v8Entries` embedding `v6Entries` is correct. The comment says it's the same as `v7` because we are replacing the index, with other indexes saying the same, however this comment is confusing because v8Entries GetWriteEntries is completely different.

I will update the comment - thanks

File: pkg/chunk/schema_util.go [hunk 9]
```diff
@@ -2,15 +2,22 @@ package chunk
...
+	"encoding/json"
 
 	"fmt"
 
 	"github.com/prometheus/common/model"
 )
 
+func metricSeriesID(m model.Metric) string {
+	h := sha256.Sum256([]byte(m.String()))
+	return string(encodeBase64Bytes(h[:]))
```
Comment (marccarre) [inline pkg/chunk/schema_util.go]: Out of curiosity, is there a special reason we use `base64` here vs., for example, `hex.EncodeToString` which:
- is more natural for such hash functions, and 
- could be more convenient to use when debugging? (or e.g. printing https://play.golang.org/p/IeHqPaaNk4)

Comment (aaron7) [inline pkg/chunk/schema_util.go]: No special reason. It's slightly more efficient to store (4 chars per 3 bytes instead of 2 chars per 2 bytes) and is seen more of an UID than a hash. Base64 is human readable and we also encode to base64 for the dynamo api (http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.LowLevelAPI.html).

File: pkg/util/merger.go [hunk 10]
```diff
@@ -2,6 +2,29 @@ package util
 
 import "github.com/prometheus/common/model"
 
+// MergeNSamples merges and dedupes n sets of already sorted sample pairs.
+func MergeNSamples(sampleSets ...[]model.SamplePair) []model.SamplePair {
+	l := len(sampleSets)
+	switch l {
+	case 0:
```
Comment (marccarre) [inline pkg/util/merger.go]: Unless I've missed it, and even though that's trivial, I don't think there is an unit test covering this `case` statement.
If I understood well, we'd need this additional test case:
```
		{
			sampleSets: [][]model.SamplePair{},
			expected:   []model.SamplePair{},
		},
```

File: pkg/util/merger.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+		result = append(result, a[i])
+	}
+	for ; j < len(b); j++ {
+		result = append(result, b[j])
+	}
+	return result
+}
+
+// MergeNSampleSets merges and dedupes n sets of already sorted sample pairs.
+func MergeNSampleSets(sampleSets ...[]model.SamplePair) []model.SamplePair {
```
Comment (marccarre) [inline pkg/util/merger.go]: Would it be worth implementing: 
- a n-way merge, instead of 
- this 2-way merge for which we then enqueue results (and repeat)?

It feels like we would re-allocate quite a few arrays and reprocess data for large number of `sampleSets`.

Comment (marccarre) [inline pkg/util/merger.go]: In case this could be useful, here is an example of doing the n-way merge using a "dedup iterator" on a "merging iterator" on a collection of "iterators": https://play.golang.org/p/5d5F7EKgZo
Lack of generics make this pretty ugly, but feedback welcome. 
The thing about it is that it is nicely composable.

File: pkg/chunk/iterator.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+// RangeValues implements the SeriesIterator interface.
+func (it *LazySeriesIterator) RangeValues(in metric.Interval) []model.SamplePair {
+	var err error
+	it.onceCreateIterator.Do(func() {
+		err = it.createSampleSeriesIterator()
+	})
+	if err != nil {
+		// TODO: Handle error.
```
Comment (marccarre) [inline pkg/chunk/iterator.go]: Is there a plan to handle this error before this PR gets merged? (I am just asking given the `TODO`)

Comment (aaron7) [inline pkg/chunk/iterator.go]: Not in this PR. I would like to discuss this - I'm unsure how to approach this.

File: pkg/querier/querier.go [hunk 13]
```diff
@@ -221,12 +204,27 @@ func (qm MergeQuerier) RemoteReadHandler(w http.ResponseWriter, r *http.Request)
...
+				OldestInclusive: from,
+				NewestInclusive: to,
+			}
+
+			// Convert iterators to matrix
+			matrix := make(model.Matrix, 0, len(iterators))
+			for _, it := range iterators {
+				ss := &model.SampleStream{
+					Metric: it.Metric().Metric,
+					Values: it.RangeValues(in),
```
Comment (marccarre) [inline pkg/querier/querier.go]: Just to point out: `RangeValues` could return `nil`. Does it matter? What do we do with these `SampleStream`s afterwards?

Comment (marccarre) [inline pkg/querier/querier.go]: It looks like we only pass these to [`ToQueryResponse`](https://github.com/weaveworks/cortex/blob/416-lazy-series-when-no-metric-name/pkg/util/compat.go#L77) but we don't check for `nil` there, so we could potentially "[panic](https://play.golang.org/p/SK_rgQtRBb)".

File: pkg/util/iterator.go [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// Metric implements the SeriesIterator interface.
+func (it SampleStreamIterator) Metric() metric.Metric {
+	return metric.Metric{Metric: it.ss.Metric}
+}
+
+// ValueAtOrBeforeTime implements the SeriesIterator interface.
+func (it SampleStreamIterator) ValueAtOrBeforeTime(ts model.Time) model.SamplePair {
+	// TODO: This is a naive inefficient approach - in reality, queries go mostly
```
Comment (marccarre) [inline pkg/util/iterator.go]: Is this something to resolve as part of this PR?

File: pkg/chunk/iterator.go (hunk 12 above)
Comment (marccarre) [inline pkg/chunk/iterator.go]: Unsure as well right now... and it looks like it could fail in many ways :slightly_frowning_face:

**Comment by aaron7** [comment]
@tomwilkie 
> Is this gtg now? I'll review this weekend if so.

Yes, it's ready for review if you get some time. Thanks :)

**Comment by tomwilkie** [comment]
Great!  I'll look at it tomorrow afternoon.

File: pkg/util/iterator.go (hunk 14 above)
Comment (aaron7) [inline pkg/util/iterator.go]: This was moved from `pkg/querier/iterator.go`. We still use `SampleStreamIterator` in some cases and the comment is still relevant. We still need improve the iterators to iterate through chunks - this PR focuses on allowing queries which do not fetch samples to execute efficiently.

File: pkg/chunk/chunk.go [hunk 15]
```diff
@@ -278,6 +279,35 @@ func (c *Chunk) decode(input []byte) error {
...
+	iterators := make([]local.SeriesIterator, 0, len(sampleStreams))
+	for _, ss := range sampleStreams {
+		iterators = append(iterators, util.NewSampleStreamIterator(ss))
+	}
+
+	return iterators, nil
+}
+
 // ChunksToMatrix converts a slice of chunks into a model.Matrix.
 func ChunksToMatrix(chunks []Chunk) (model.Matrix, error) {
```
Comment (tomwilkie) [inline pkg/chunk/chunk.go]: Only used in tests now, could be moved into them.

File: pkg/chunk/chunk_store.go [hunk 16]
```diff
@@ -150,18 +151,35 @@ func (c *Store) calculateDynamoWrites(userID string, chunks []Chunk) (WriteBatch
...
+	}
+
+	// Otherwise we will create lazy iterators for all series in our index
+	return c.getSeriesIterators(ctx, from, through, matchers, metricNameMatcher)
+}
 
-	// Fetch chunk descriptors (just ID really) from storage
-	chunks, err := c.lookupChunksByMatchers(ctx, from, through, matchers)
+func (c *Store) getMetricNameIterators(ctx context.Context, from, through model.Time, allMatchers []*metric.LabelMatcher, metricName model.LabelValue) ([]local.SeriesIterator, error) {
+	chunks, err := c.getMetricNameChunks(ctx, from, through, allMatchers, metricName)
```
Comment (tomwilkie) [inline pkg/chunk/chunk_store.go]: My preference would be to move the contents of `getMetricNameChunks` into here, as this function seems overly short.  I see that it is used from the tests, so you'd need to do some work there, in which case it might not be worth the effort.

File: pkg/chunk/iterator.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+	return (*it.sampleSeriesIterator).RangeValues(in)
+}
+
+// Close implements the SeriesIterator interface.
+func (it *LazySeriesIterator) Close() {}
+
+func (it *LazySeriesIterator) createSampleSeriesIterator() error {
+	metricName, ok := it.metric[model.MetricNameLabel]
+	if !ok {
+		return fmt.Errorf("series does not have a metric name")
```
Comment (tomwilkie) [inline pkg/chunk/iterator.go]: Can you push this error into the creation of the LazySeriesIterator?

File: pkg/chunk/iterator.go [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	ctx := context.Background()
```
Comment (tomwilkie) [inline pkg/chunk/iterator.go]: Can you open a ticket upstream to add contexts and error returns to iterators.  Will probably have to wait until post v2 though.

File: pkg/chunk/iterator.go [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+	sampleSeriesIterators, err := it.chunkStore.getMetricNameIterators(ctx, it.from, it.through, it.matchers, metricName)
+	if err != nil {
+		return err
+	}
+
+	// We should only expect one sampleSeriesIterator because we are dealing
+	// with one series.
```
Comment (tomwilkie) [inline pkg/chunk/iterator.go]: This isn't correct - the combination of metric + name and matcher might not necessarily produce a single series.  Image I do the query `{}` - matchers will be an empty list, and metric name will be `foo`.  `foo{}` will mach many series.

The correct thing to do here is to ignore `matchers` (shouldn't even be passing it to the iterators constructor), and instead create your own matchers based on the metric.  Each label=value pair should become a label=value equality matcher. Then it should guarantee a single series.

File: pkg/util/merger.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+	c := make(chan []model.SamplePair, l)
+	for _, ss := range sampleSets {
+		c <- ss
+	}
+	for ; l > 1; l-- {
+		left, right := <-c, <-c
+		c <- MergeSampleSets(left, right)
+	}
+	return <-c
```
Comment (tomwilkie) [inline pkg/util/merger.go]: Use of the channel seems a bit gratuitous here.  Wouldn't a slice work just as well?  Or even the recursion approach? I guess its not a big deal.

**Comment by tomwilkie** [comment]
@aaron7 sorry took so long to review.  Mainly minor nits, but one mistake in how you materialise a lazy iterator that needs fixing.

**Comment by aaron7** [comment]
> @aaron7 sorry took so long to review. Mainly minor nits, but one mistake in how you materialise a lazy iterator that needs fixing.

Thanks @tomwilkie - I really appreciate the review. I will make the changes today and get myself onto irc.

File: pkg/querier/querier.go (hunk 13 above)
Comment (aaron7) [inline pkg/querier/querier.go]: I've had a look into this and it doesn't look like it will panic because we are ranging over a `nil` slice which does 0 iterations. https://github.com/golang/go/blob/master/doc/go_spec.html#L5011
https://play.golang.org/p/4vvbbHMKUo

File: pkg/chunk/chunk_store.go (hunk 16 above)
Comment (aaron7) [inline pkg/chunk/chunk_store.go]: Yep, once `getMetricNameIterators` converts the chunks from `getMetricNameChunks` into iterators, we cannot access the chunks for testing through the iterator interface. We would like access to these to test whether the right chunks are fetched or not.

Comment (aaron7) [inline pkg/chunk/chunk_store.go]: We can't test at the interval level as we are concerned about which chunks are fetched to create the iterators. We could expose the chunks, but it's a prometheus interface and it doesn't seem like the right thing to do, exposing something that is not related. This currently seems like the best way to get around testing constraints.

File: pkg/util/merger.go (hunk 20 above)
Comment (aaron7) [inline pkg/util/merger.go]: Agreed, will go with the recursion approach here.

File: pkg/util/merger.go (hunk 11 above)
Comment (aaron7) [inline pkg/util/merger.go]: As discussed offline, we are going to keep with the merge approach for now in this PR because the most common case is only a few sample sets and the code is simpler.

File: pkg/ingester/ingester_test.go [hunk 21]
```diff
@@ -82,6 +82,36 @@ func matrixToSamples(m model.Matrix) []model.Sample {
 	return samples
 }
 
+// chunksToMatrix converts a slice of chunks into a model.Matrix.
+func chunksToMatrix(chunks []chunk.Chunk) (model.Matrix, error) {
```
Comment (tomwilkie) [inline pkg/ingester/ingester_test.go]: You could still have this as a method on Chunk, even if its in a _test.go.  But no worries either way.

File: pkg/chunk/chunk.go [hunk 22]
```diff
@@ -308,37 +308,8 @@ func chunksToIterators(chunks []Chunk) ([]local.SeriesIterator, error) {
...
-	for _, ss := range sampleStreams {
-		matrix = append(matrix, ss)
-	}
-
-	return matrix, nil
-}
-
-func (c *Chunk) samples() ([]model.SamplePair, error) {
+// Samples returns all SamplePairs for the chunk.
+func (c *Chunk) Samples() ([]model.SamplePair, error) {
```
Comment (tomwilkie) [inline pkg/chunk/chunk.go]: This doesn't need to be made public - tests are part of the same module.

File: pkg/chunk/iterator.go [hunk 23]
```diff
@@ -29,14 +29,18 @@ type LazySeriesIterator struct {
 }
 
 // NewLazySeriesIterator creates a LazySeriesIterator.
-func NewLazySeriesIterator(chunkStore *Store, metric model.Metric, from model.Time, through model.Time, matchers []*metric.LabelMatcher) *LazySeriesIterator {
+func NewLazySeriesIterator(chunkStore *Store, metric model.Metric, from model.Time, through model.Time, matchers []*metric.LabelMatcher) (*LazySeriesIterator, error) {
+	_, ok := metric[model.MetricNameLabel]
```
Comment (tomwilkie) [inline pkg/chunk/iterator.go]: Cache the metric name in the struct and remove the error case in `createSampleSeriesIterator`?

**Comment by tomwilkie** [comment]
Sorry took so long to have another look.  Couple of nits but otherwise LGTM.  Rebase and merge!

File: pkg/chunk/chunk.go (hunk 22 above)
Comment (aaron7) [inline pkg/chunk/chunk.go]: This was made public because `chunksToMatrix` was moved to the `ingester` package which used `.samples()`. However I can move `ChunksToMatrix` back to the chunk package and expose it there as suggested.

File: pkg/ingester/ingester_test.go (hunk 21 above)
Comment (aaron7) [inline pkg/ingester/ingester_test.go]: See comment below. Will move back to chunk package.

File: pkg/chunk/iterator.go (hunk 23 above)
Comment (aaron7) [inline pkg/chunk/iterator.go]: good idea, will update

File: pkg/ingester/ingester_test.go (hunk 21 above)
Comment (tomwilkie) [inline pkg/ingester/ingester_test.go]: Nah don't bother, I missed the package move.

**Comment by aaron7** [comment]
Rebased - will test before merging.

**Comment by aaron7** [comment]
Tested locally with `http_request_duration_microseconds`, `up` and `count({__name__=~".+"}) by (__name__)`. 

We're seeing `Error: QueryPages error: table=cortex, err=SerializationError: failed decoding JSON RPC response caused by: header line too long` errors locally for the fuzzy queries when we have a few chunks in DynamoDB, however we believe this could be a problem with localstack. The agreed plan is to roll out to dev and test against DynamoDB and file an issue with localstack if it is inconsistent.

I will also be adding some more metrics to get information about how many pages we are seeing for each query.

**Comment by bboreham** [comment]
> It will replace the metric name index

how?
